cpp/include/raft/core/resource/multi_gpu.hpp

```diff
@@ -0,0 +1,134 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/resource/device_id.hpp>
+#include <raft/core/resource/resource_types.hpp>
+#include <raft/core/resources.hpp>
+
+#include <memory>
+
+namespace raft::resource {
+
+class multi_gpu_resource : public resource {
+ public:
+  multi_gpu_resource() : world_resources_(std::make_unique<std::vector<raft::resources>>(0)) {}
+  ~multi_gpu_resource() override {}
+  void* get_resource() override { return world_resources_.get(); }
+
+ private:
+  std::unique_ptr<std::vector<raft::resources>> world_resources_;
+};
+
+/** Factory that knows how to construct a specific raft::resource to populate the res_t. */
+class multi_gpu_resource_factory : public resource_factory {
+ public:
+  resource_type get_resource_type() override { return resource_type::MULTI_GPU; }
+  resource* make_resource() override { return new multi_gpu_resource(); }
+};
+
+class main_gpu_resource : public device_id_resource {
+ public:
+  main_gpu_resource() : device_id_resource() {}
+  ~main_gpu_resource() override {}
+};
+
+class main_gpu_resource_factory : public resource_factory {
+ public:
+  resource_type get_resource_type() override { return resource_type::MAIN_GPU_ID; }
+  resource* make_resource() override { return new main_gpu_resource(); }
+};
+
+inline int& get_main_gpu_id(resources const& res)
+{
+  if (!res.has_resource_factory(resource_type::MAIN_GPU_ID)) {
+    res.add_resource_factory(std::make_shared<main_gpu_resource_factory>());
+  }
+  return *res.get_resource<int>(resource_type::MAIN_GPU_ID);
+};
+
+/**
+ * Retrieves a multi gpu resource from raft res if it exists, otherwise initializes it and returns
+ * it.
+ *
+ * @param[in] res the raft resources object
+ * @return resource for each device in multi-gpu world
+ */
+inline std::vector<raft::resources>& get_multi_gpu_resource(resources const& res)
+{
+  if (!res.has_resource_factory(resource_type::MULTI_GPU)) {
+    res.add_resource_factory(std::make_shared<multi_gpu_resource_factory>());
+  }
+  return *res.get_resource<std::vector<raft::resources>>(resource_type::MULTI_GPU);
+};
+
+/**
+ * @brief Returns true if res has a multi GPU resource type
+ */
+inline bool is_multi_gpu(resources const& res)
+{
+  return res.has_resource_factory(resource_type::MULTI_GPU);
+};
+
+/**
+ * @brief Get number of gpus in multi-gpu world
+ */
+inline int get_world_size(resources const& res)
+{
+  return raft::resource::get_multi_gpu_resource(res).size();
+}
+
+/**
+ * @brief Get specific GPU's raft::resources object
+ */
+inline const raft::resources& get_device_resources_for_gpu_id(resources const& res, int gpu_id)
+{
+  std::vector<raft::resources>& world_resources = raft::resource::get_multi_gpu_resource(res);
+  return world_resources[gpu_id];
```

Comment:
It is very important to make a distinction between a rank and a device ID. Here `world_resources[gpu_id]` would effectively access the `device_resource` for the rank number `gpu_id`.

Reply:
Depending on how we are using the term "gpu_id" here. Are we saying the id for a specific GPU? Those are allowed to skip numbers. A `rank` (or what I also call `canonical id` is not allowed to skip numbers. I think this is the differentiation that @viclafargue is referring to.

Reply:
It is being used more like a rank. (e.g. 0,1,2,3 when we actually use GPUs 2,4,6,7) but I agree that it might be confusing.

Reply:
What was previously referred to as the root rank has to be part of the set of device IDs. The function formerly named `set_current_device_to_root_rank` (now renamed to `set_current_device_to_main_gpu`) searched for the root rank to return its `device_resource`. For example, consider a clique with GPU IDs {2, 4, 6, 7} Previously the root rank defaulted to 0, and the merge operation was performed on GPU 2. Now, if the `device_resource_mnmg_nccl` class is instantiated on GPU 0, the ANN merge operation will be executed on GPU 2. However, if it is instantiated on say GPU 8, a segmentation fault will occur.